File: answers/other/in sap/Path Sum II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        
        path_list = []
        self.dfs(root, [], sum, path_list)
        return path_list

    def dfs(self, root, path, sum, path_list):
    	if root:
    		if None == root.left and None == root.right and root.val == sum:
    			path_list.append(path + [root.val])
    		self.dfs(root.left, path + [root.val], sum - root.val, path_list)
    		self.dfs(root.right, path + [root.val], sum - root.val, path_list)

File: answers/leetcode/Path Sum II/Path Sum II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<>();
        this.dfs(new ArrayList<>(), root, sum, result);
        return result;
    }

    private void dfs(List<Integer> path, TreeNode root, int sum, List<List<Integer>> result) {
        if (null == root) {
            return;
        }
        path.add(root.val);
        if (null == root.left && null == root.right && sum == root.val) {
            result.add(new ArrayList<>(path));
        }
        this.dfs(path, root.left, sum - root.val, result);
        this.dfs(path, root.right, sum - root.val, result);
        path.remove(path.size() - 1);
    }
}